File: ankor-core/src/main/java/at/irian/ankor/change/Change.java
package at.irian.ankor.change;

import at.irian.ankor.messaging.AnkorIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;

import java.util.Collection;

/**
 * @author Manfred Geiler
 */
public class Change {

    /**
     * Type of this change.
     */
    private final ChangeType type;

    /**
     * (Optional) key or index of this change for change types {@link ChangeType#insert},
     * {@link ChangeType#delete}, and {@link ChangeType#replace}.
     */
    private final Object key;

    /**
     * Value of this change.
     */
    @JsonInclude(JsonInclude.Include.ALWAYS)
    private final Object value;

    protected Change(ChangeType type, Object key, Object value) {
        this.type = type;
        this.key = key;
        this.value = value;
    }

    public static Change valueChange(Object newValue) {
        return new Change(ChangeType.value, null, newValue);
    }

    public static Change insertChange(int idx, Object insertedElement) {
        return new Change(ChangeType.insert, idx, insertedElement);
    }

    public static Change deleteChange(Object key) {
        return new Change(ChangeType.delete, key, null);
    }

    public static Change replaceChange(int fromIdx, Collection newElements) {
        return new Change(ChangeType.replace, fromIdx, newElements);
    }

    public Change withValue(Object newValue) {
        return new Change(type, key, newValue);
    }

    public ChangeType getType() {
        return type;
    }

    public Object getKey() {
        return key;
    }

    public Object getValue() {
        return value;
    }


    @AnkorIgnore
    public int getKeyAsIndex() {
        if (key instanceof Number) {
            return ((Number) key).intValue();
        } else if (key instanceof String) {
            return Integer.parseInt((String) key);
        } else {
            throw new IllegalArgumentException("list/array index of type " + key.getClass());
        }
    }

    @AnkorIgnore
    public String getKeyAsMapKey() {
        if (key instanceof String) {
            return (String)key;
        } else {
            throw new IllegalArgumentException("map key of type " + key.getClass());
        }
    }



    @SuppressWarnings("RedundantIfStatement")
    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Change)) {
            return false;
        }

        Change change = (Change) o;

        if (key != null ? !key.equals(change.key) : change.key != null) {
            return false;
        }
        if (type != change.type) {
            return false;
        }
        if (value != null ? !value.equals(change.value) : change.value != null) {
            return false;
        }

        return true;
    }

    @Override
    public int hashCode() {
        int result = type.hashCode();
        result = 31 * result + (key != null ? key.hashCode() : 0);
        result = 31 * result + (value != null ? value.hashCode() : 0);
        return result;
    }

    @Override
    public String toString() {
        return "Change{" +
               "type=" + type +
               ", key=" + key +
               ", value=" + value +
               "}";
    }


}

File: #ANKOR.net/Ankor.Core/Change/Change.cs
﻿using System.Collections;

namespace Ankor.Core.Change {

	public class Change {
		public ChangeType Type { get; private set; }
		public object Key { get; private set; }
		public object Value { get; private set; }

		// for json serializing
		private Change() {}

		private Change(ChangeType type, object key, object value) {
			this.Type = type;
			this.Key = key;
			this.Value = value;
		}

		public static Change ValueChange(object val) {
			return new Change(ChangeType.Value, null, val);
		}

		public static Change InsertChange(long idx, object insertedElement) {
				return new Change(ChangeType.Insert, idx, insertedElement);
		}

		public static Change DeleteChange(object key) {
				return new Change(ChangeType.Delete, key, null);
		}

		public static Change ReplaceChange(long fromIdx, ICollection newElements) {
			return new Change(ChangeType.Replace, fromIdx, newElements);
		}

		protected bool Equals(Change other) {
			return Type.Equals(other.Type) && Equals(Key, other.Key) && Equals(Value, other.Value);
		}

		public static bool operator ==(Change left, Change right) {
			return Equals(left, right);
		}

		public static bool operator !=(Change left, Change right) {
			return !Equals(left, right);
		}

		public override string ToString() {
			return string.Format("Type: {0}, Key: {1}, Value: {2}", Type, Key, Value);
		}

		public override bool Equals(object obj) {
			if (ReferenceEquals(null, obj)) {
				return false;
			}
			if (ReferenceEquals(this, obj)) {
				return true;
			}
			if (obj.GetType() != this.GetType()) {
				return false;
			}
			return Equals((Change) obj);
		}

		public override int GetHashCode() {
			unchecked {
				int hashCode = Type.GetHashCode();
				hashCode = (hashCode*397) ^ (Key != null ? Key.GetHashCode() : 0);
				hashCode = (hashCode*397) ^ (Value != null ? Value.GetHashCode() : 0);
				return hashCode;
			}
		}

	}

	public enum ChangeType {
		Value,
		Insert,
		Delete,
		Replace
	}
}
